File: packages/python-sdk/src/blindfold/regex/detectors/us/ssn.py
"""US Social Security Number detector"""

import re
from typing import List

from ...base import RegexDetector
from ...entities import EntityType, PIIMatch
from ...registry import register_region
from ...validators import ssn_valid_format

# SSN with separators (dashes, spaces, or dots)
_SSN_WITH_SEP = re.compile(
    r"\b(?!000|666|9\d{2})\d{3}"
    r"[-\s.]"
    r"(?!00)\d{2}"
    r"[-\s.]"
    r"(?!0000)\d{4}\b"
)

# SSN without separators (9 bare digits) — requires context to avoid noise
_SSN_NO_SEP = re.compile(
    r"\b(?!000|666|9\d{2})\d{3}"
    r"(?!00)\d{2}"
    r"(?!0000)\d{4}\b"
)


@register_region("us")
class SsnDetector(RegexDetector):
    entity_type = EntityType.SSN
    score = 0.85
    context_keywords = [
        "ssn", "social security", "social sec", "ss#", "ss #",
        "social", "tax id", "taxpayer", "tin",
        "social number", "soc number", "socialnum",
    ]
    context_required = False
    context_window = 100

    # Primary pattern (with separators) — used by base class pre_check
    pattern = _SSN_WITH_SEP
    validator = staticmethod(ssn_valid_format)

    def iter_matches(self, text: str) -> List[PIIMatch]:
        results: List[PIIMatch] = []

        # Branch 1: SSN with separators (no context required)
        for match in _SSN_WITH_SEP.finditer(text):
            matched_text = match.group()
            if not ssn_valid_format(matched_text):
                continue
            has_ctx = self._has_context(text, match.start(), match.end())
            score = 1.0 if has_ctx else self.score
            results.append(PIIMatch(
                entity_type=self.entity_type.value,
                text=matched_text,
                start=match.start(),
                end=match.end(),
                score=score,
            ))

        # Branch 2: SSN without separators (context REQUIRED)
        for match in _SSN_NO_SEP.finditer(text):
            matched_text = match.group()
            start = match.start()
            end = match.end()
            # Skip if this overlaps with an already-found separated SSN
            if any(r.start <= start < r.end for r in results):
                continue
            if not ssn_valid_format(matched_text):
                continue
            if not self._has_context(text, start, end):
                continue
            results.append(PIIMatch(
                entity_type=self.entity_type.value,
                text=matched_text,
                start=start,
                end=end,
                score=1.0,
            ))

        return results

File: packages/js-sdk/src/regex/detectors/us/ssn.ts
// US Social Security Number detector

import { Detector } from '../../base'
import { EntityType, PIIMatch } from '../../entities'
import { registerRegion } from '../../registry'
import { ssnValidFormat } from '../../validators'

// SSN with separators (dashes, spaces, or dots)
const SSN_WITH_SEP = new RegExp(
  '\\b(?!000|666|9\\d{2})\\d{3}[-.\\s](?!00)\\d{2}[-.\\s](?!0000)\\d{4}\\b',
  'g'
)

// SSN without separators (9 bare digits) — requires context to avoid noise
const SSN_NO_SEP = new RegExp(
  '\\b(?!000|666|9\\d{2})\\d{3}(?!00)\\d{2}(?!0000)\\d{4}\\b',
  'g'
)

class SsnDetector extends Detector {
  entityType = EntityType.SSN
  score = 0.85
  contextKeywords = [
    'ssn', 'social security', 'social sec', 'ss#', 'ss #',
    'social', 'tax id', 'taxpayer', 'tin',
    'social number', 'soc number', 'socialnum',
  ]
  contextRequired = false
  contextWindow = 100

  iterMatches(text: string): PIIMatch[] {
    const results: PIIMatch[] = []

    // Branch 1: SSN with separators (no context required)
    SSN_WITH_SEP.lastIndex = 0
    let match: RegExpExecArray | null
    while ((match = SSN_WITH_SEP.exec(text)) !== null) {
      const matchedText = match[0]
      if (!ssnValidFormat(matchedText)) continue
      const start = match.index
      const end = start + matchedText.length
      const hasCtx = this.hasContext(text, start, end)
      results.push({
        entityType: this.entityType,
        text: matchedText,
        start,
        end,
        score: hasCtx ? 1.0 : this.score,
      })
    }

    // Branch 2: SSN without separators (context REQUIRED)
    SSN_NO_SEP.lastIndex = 0
    while ((match = SSN_NO_SEP.exec(text)) !== null) {
      const matchedText = match[0]
      const start = match.index
      const end = start + matchedText.length
      // Skip if this overlaps with an already-found separated SSN
      if (results.some((r) => r.start <= start && start < r.end)) continue
      if (!ssnValidFormat(matchedText)) continue
      if (!this.hasContext(text, start, end)) continue
      results.push({
        entityType: this.entityType,
        text: matchedText,
        start,
        end,
        score: 1.0,
      })
    }

    return results
  }
}

registerRegion('us', SsnDetector)

export { SsnDetector }
